byzcoin/statetrie.go

```diff
@@ -149,13 +155,84 @@ func (t *stagingStateTrie) GetVersion() Version {
 	return readVersion(t)
 }
 
+func (t *stagingStateTrie) LoadConfigFromTrie() (*ChainConfig, error) {
+	return t.loadConfigFromTrie(t)
+}
+
+func (t *stagingStateTrie) LoadDarcFromTrie(id darc.ID) (*darc.Darc, error) {
+	return t.loadDarcFromTrie(t, id)
+}
+
+type trieCache struct {
+	config *ChainConfig
+	darcs  map[string]*darc.Darc
+}
+
+func (tc *trieCache) loadConfigFromTrie(st ReadOnlyStateTrie) (*ChainConfig, error) {
+	if tc.config != nil {
+		return tc.config, nil
+	}
+
+	// Find the genesis-darc ID.
+	val, _, contract, _, err := GetValueContract(st, NewInstanceID(nil).Slice())
+	if err != nil {
+		return nil, xerrors.Errorf("reading trie: %w", err)
+	}
+	if string(contract) != ContractConfigID {
+		return nil, xerrors.New("did not get " + ContractConfigID)
+	}
+
+	tc.config = &ChainConfig{}
+	err = protobuf.DecodeWithConstructors(val, tc.config,
+		network.DefaultConstructors(cothority.Suite))
+	if err != nil {
+		return nil, xerrors.Errorf("decoding config: %v", err)
+	}
+
+	return tc.config, nil
+}
+
+func (tc *trieCache) loadDarcFromTrie(st ReadOnlyStateTrie,
+	id darc.ID) (*darc.Darc, error) {
+	if len(tc.darcs) > 0 {
+		if d := tc.darcs[string(id)]; d != nil {
+			return d, nil
+		}
+	} else {
+		tc.darcs = make(map[string]*darc.Darc)
+	}
+	darcBuf, _, contract, _, err := st.GetValues(id)
+	if err != nil {
+		return nil, xerrors.Errorf("reading trie: %v", err)
+	}
+	config, err := tc.loadConfigFromTrie(st)
+	if err != nil {
+		return nil, xerrors.Errorf("reading trie: %v", err)
+	}
+	var ok bool
+	for _, id := range config.DarcContractIDs {
+		if contract == id {
+			ok = true
+		}
+	}
+	if !ok {
+		return nil, xerrors.New("the contract \"" + contract + "\" is not in the set of DARC contracts")
+	}
+	tc.darcs[string(id)], err = darc.NewFromProtobuf(darcBuf)
+	if err != nil {
+		return nil, xerrors.Errorf("decoding darc: %v", err)
```

Comment:
Can help with knowing what was wrong in the encoded DARC (which key is missing, what failed exactly, ASO)
```suggestion
		return nil, fmt.Errorf("decoding darc: %w", err)
```

Reply:
I thought that if you pass `%v` with an `err`, it will correctly create a stack-trace, no?

Reply:
Not from what I read on [godoc](https://golang.org/pkg/fmt/#Errorf)

Reply:
The `Unwrap` method is used to be able to check if an error equals to an existing error. Which is not what we want to do here. So I just follow the DEDIS standard here:

```
4. Errors are systematically wrapped with a meaningful message using xerrors.Errorf and the %v verb.
```